### ISSUE TITLE
Backend module generating unwanted documentation despite `@moduledoc false`

### ISSUE THREAD

**Issue by bobwaycott** [kipcole9/cldr#113]
Thanks for the fantastic `Cldr` lib & its related packages. They're incredible.

I am encountering some odd behavior with my `Cldr` backend module and unwanted documentation generation.

Here's the backend module def:

```elixir
# lib/myapp/cldr.ex
defmodule MyApp.Cldr do
  @moduledoc false
  use Cldr,
    default_locale: MyApp.Config.default_locale(),
    locales: MyApp.Config.locales(),
    json_library: Jason,
    providers: [Cldr.Currency, Cldr.Number]
end
```

**Fwiw:** I have actually defined this backend for `ex_money`, as it says a `default_cldr_backend` is required. My config that points to my custom backend looks like this:

```elixir
config :ex_money,
  auto_start_exchange_rate_service: false,
  default_cldr_backend: MyApp.Cldr,
  json_library: Jason
```

Despite the `@moduledoc false` declaration, my library docs keep generating a bunch of `Cldr`-related doc pages:

<img width="317" alt="unexpected-docs" src="https://user-images.githubusercontent.com/196710/54294589-f2ec8880-4587-11e9-930b-8f9cec28bbc1.png">

Additionally, I'm a bit confused as to why I keep seeing this message:

```shell
Generating MyApp.Cldr for 3 locales named ["en", "en-001", "root"] with a default locale named "en"
warning: redefining module MyApp.Cldr.Currency (current version loaded from _build/dev/lib/myapp/ebin/Elixir.MyApp.Cldr.Currency.beam)
  lib/myapp/cldr.ex:1
```

I use a lot of `use` statements throughout my library for functionality. Usually, no docs are generated when a `@moduledoc false` exists. However, I cannot seem to figure out why `Cldr`-related modules are generating unwanted docs despite the directive explicitly stating not to do so.

Am I missing something really obvious here, or are `Cldr` modules misbehaving and creating docs even when they aren't wanted?

**Comment by kipcole9** [comment]
@bobwaycott thanks for the feedback, it's appreciated.  I have gone out of my way to make sure there are (good I hope) docs for the generated modules because they are intended as the primary API for the CLDR* family of libraries.  But I can also understand that this is not a concern of an `ex_money` consumer.

The reason you see the docs is that your `@moduledoc false` is followed, in the generated code, by `@moduledoc .....`.

I can make the docs optional and will do so (its going to take a few days thought).  I did argue with myself over this topic early on in development :-)

The `Generating MyApp.Cldr for 3 locales named ["en", "en-001", "root"] with a default locale named "en"` message is generated for each an any backend module that is defined. Its actually the backend module that holds all the data that drives `ex_money`. It has the formatting data, the currency symbols, the rounding rules and so on.  I appreciate that maybe you only use the "en" locale but there are other use cases that use more.  The reason for the `:default_backend` requirement is so that this data is available without it having to be a required parameter on each function call (it is an option though if required).

The `warning: redefining module MyApp.Cldr.Currency (current version loaded from _build/dev/lib/myapp/ebin/Elixir.MyApp.Cldr.Currency.beam)
  lib/myapp/cldr.ex:1` bugs the heck out of me so I'll dig into that - its lingered around for too long so I;ll get that sorted.

In summary, my plan is to:
1. Make docs in generated modules optional
2, Keep the `Generating MyApp.Cldr for 3 locales named ["en", "en-001", "root"] ...` message
3. Get rid of the `warning: redefining module MyApp.Cldr.Currency...` message

Happy to take feedback and thanks much for the encouragement on this whole project that has somehow not got a life of its own!

**Comment by bobwaycott** [comment]
@kipcole9 Hey, thanks for the detailed reply. Really helps understand your thinking and approach. I had a feeling that might be what was happening. Also, the CLDR family of docs have been top-notch. I'm a heavy documenter myself, so it's been refreshing to see such thorough docs on all the things.

I'm not sure if it'd work for the CLDR libraries, but in case it does, I thought I'd share the way I approach handling the same kind of issue in my own library code. I, too, go to great lengths to provide a bunch of documentation that can be generated for anything that brings a module in via `use`, while also needing at times to let the documentation be ignored. I've been handling it like this:

```elixir
defmodule MyLib.SomeModule do
  @moduledoc false

  @doc false
  defmacro __using__(opts \\ []) do
    quote bind_quoted: [opts: opts] do
      @mod_opts opts

      if is_nil(Module.get_attribute(__MODULE__, :moduledoc)) do
        @moduledoc """
        A bunch of detailed documentation about all the things that will
        be generated with `mix docs` unless someone explicitly adds a
        `@moduledoc` directive to the module that is using this one.
        """
      end
  end
end
```

For my needs, this has been serving me well for a bunch of modules that I intend to only ever be `use`d and not interacted with directly, so it obviously may not be a perfect fit for CLDR's approach. I don't want to make it harder on you to keep fantastic documentation happening. My approach has been working because with the library I've been building out, I don't want these underlying modules to be documented themselves (hence the opening `@moduledoc false`), but I have a bunch of functionality defined inside the `defmacro __using__` block--both the `@moduledoc` nil test & functions with their own `@doc` directives that have a bunch of dynamic things happening to generate docs that match up to the module that has included the `use` statement.

Anyway, the simple result is that docs wind up generating **only** if there is no `@moduledoc` directive at all. If `@moduledoc` is `false`, there are no docs, otherwise a `@moduledoc` being provided let's the using module have its own module docstring, and then the rest of the function docs are brought in from the used module. Hope that makes some kind of sense and might be even slightly useful (though I'm sure you've already got your own good ideas).

Thanks again for such a fantastic library, and all the great docs.

**Comment by kipcole9** [comment]
Appreciate the feeback, thanks.

I have as of [this commit](https://github.com/kipcole9/cldr/commit/0be7a7db9a5e1e51f1208903106148d7c47f4f62) made doc generation configurable for [ex_cldr](https://github.com/kipcole9/cldr).  I am yet to do [ex_cldr_numbers](https://github.com/kipcole9/cldr_numbers) or [ex_cldr_currencies](https://github.com/kipcole9/cldr_currencies) both of which are also dependencies of [ex_money](https://github.com/kipcole9/money).  I'm on a looooong flight today and will update them as well.

Since there are potentially many different modules generated depending on what providers are configured I have taken a slightly different approach as follows (from the changelog):

### Enhancements

* Makes generation of documentation for backend modules options.  This is implemented with the `:generate_docs` option to the backend configuration.  The default is `true`. For example:

```
defmodule MyApp.Cldr do
  use Cldr,
    default_locale: "en-001",
    locales: ["en", "ja"],
    gettext: MyApp.Gettext,
    generate_docs: false
end
```

**Comment by bobwaycott** [comment]
Sounds like the easiest way to do it for sure. I'll keep an eye on how it comes together. Safe travels!

**Comment by kipcole9** [comment]
Updated all the `ex_cldr_*` libs to respect the `generate_docs: false` configuration option. Please:
```
mix deps.update ex_cldr ex_cldr_numbers ex_cldr_currencies
```
Then add to your backend configuration:
```
defmodule MyApp.Cldr do
  use Cldr,
    ...
    generate_docs: false

end
```
And you should be good to go. Thanks for the report and the motivation to add this option.

**Comment by bobwaycott** [comment]
Hey @kipcole9 ! Thanks a bunch. Easy upgrade. Looks like there are still two items generating in my docs right now:

- `Cldr.Rbnf.Ordinal`
- `Cldr.Rbnf.Spellout`

Should those be ignored by these changes, too?

**Comment by kipcole9** [comment]
Sorry @bobwaycott, I missed those two,  Fixed now in `ex_cldr_numbers` version 2.4.4. Please:
```
mix deps.update ex_cldr_numbers
```
and you should be good to go.

**Comment by bobwaycott** [comment]
Awesome! Thanks again, @kipcole9
